File: Stacks/10.Min Stack.ts
class MinStack {
    private stack: [number, number][]; // Each element is a tuple [value, currentMin]

    constructor() {
        this.stack = [];
    }

    /**
     * Pushes a value onto the stack.
     * @param val - The value to push.
     */
    push(val: number): void {
        if (!this.stack.length) {
            this.stack.push([val, val]); // If the stack is empty, the value is also the current minimum
        } else {
            const currentMin = this.stack[this.stack.length - 1][1];
            this.stack.push([val, Math.min(val, currentMin)]); // Push the value and the new minimum
        }
    }

    /**
     * Removes the top element from the stack.
     */
    pop(): void {
        this.stack.pop();
    }

    /**
     * Returns the top element of the stack.
     * @returns The top element.
     */
    top(): number {
        return this.stack[this.stack.length - 1][0];
    }

    /**
     * Returns the minimum value in the stack.
     * @returns The minimum value.
     */
    getMin(): number {
        return this.stack[this.stack.length - 1][1];
    }
}

// Usage example:
// const obj = new MinStack();
// obj.push(val);
// obj.pop();
// const topValue = obj.top();
// const minValue = obj.getMin();

File: Stacks/10.Min Stack.cpp
#include <stack>
#include <algorithm>
#include <optional>
#include <stdexcept>

class MinStack {
private:
    // Stack stores pairs of (value, current minimum)
    std::stack<std::pair<int, int>> st;

public:
    // Constructor (explicitly defined for clarity)
    MinStack() = default;

    // Push a value onto the stack
    void push(int val) {
        if (st.empty()) {
            st.push({val, val}); // First element, min is itself
        } else {
            int currentMin = st.top().second;
            st.push({val, std::min(val, currentMin)}); // Update current minimum
        }
    }

    // Pop the top element from the stack
    void pop() {
        if (st.empty()) {
            throw std::runtime_error("Cannot pop from an empty stack.");
        }
        st.pop();
    }

    // Get the top element of the stack
    int top() const {
        if (st.empty()) {
            throw std::runtime_error("Stack is empty. No top element.");
        }
        return st.top().first;
    }

    // Get the minimum element in the stack
    int getMin() const {
        if (st.empty()) {
            throw std::runtime_error("Stack is empty. No minimum element.");
        }
        return st.top().second;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj;
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */
